### PR TITLE
Add official Dockerfile

### PR DESCRIPTION
**Pull request by pm47**
Follow-up to #1.

I think building native binaries on docker is not the right approach. It's better to run on the jvm: easier build, faster runtime and compatible with linux arm64, which we don't support otherwise.

TODO:
- [x] remove `lightning-kmp` build once https://github.com/ACINQ/lightning-kmp/pull/622 is merged
- [x] remove `RUN mkdir .phoenix` once #15 is merged

cc @sethforprivacy

### CODE REVIEW

**Comment by sethforprivacy** [comment]
Will take a deeper look asap, but at first glance I love this approach!

**Comment by tsmith123** [comment]
I'm getting "ERROR: failed to solve: eclipse-temurin:17-jre-alpine: no match for platform in manifest: not found" running on an M2 Macbook

**Comment by pm47** [comment]
Try with eclipse-temurin:21-jre-alpine

**Comment by tsmith123** [comment]
Changed both to 21 and it seems better but now I'm getting:

"ERROR: failed to solve: process "/bin/sh -c git clone --recursive --branch ${LIGHTNING_KMP_BRANCH}     https://github.com/ACINQ/lightning-kmp .     && test `git rev-parse HEAD` = ${LIGHTNING_KMP_COMMIT_HASH} || exit 1     && ./gradlew publishToMavenLocal -x dokkaHtml" did not complete successfully: exit code: 1"

**Comment by pm47** [comment]
Rebased on master. 

> Changed both to 21 and it seems better but now I'm getting:
> "ERROR: failed to solve: process "/bin/sh -c git clone --recursive --branch ${LIGHTNING_KMP_BRANCH} https://github.com/ACINQ/lightning-kmp . && test git rev-parse HEAD = ${LIGHTNING_KMP_COMMIT_HASH} || exit 1 && ./gradlew publishToMavenLocal -x dokkaHtml" did not complete successfully: exit code: 1"

@tsmith123 your issue should be fixed at 5630fa0.

**Review by sstone — COMMENTED** [review]
Building this docker image  still won't work on macOS arm64 because of https://github.com/gradle/gradle/issues/24875.
It could be fixed by using a glibc-based bistro (ubuntu, ...) or maybe switching to a snapshot version of gradle 8.8 but I'm not sure it's worth it as macOS arm64 is not a production target (and we  already publish a native `phoenixd` exe for macOS arm64).

**Comment by pm47** [comment]
So @sstone should work on all 4 combinations linux/macos x64/arm64?

More portability is better for docker I think. Besides, ubuntu 21 + jdk21 seems tailor made for a bitcoin app 😄

**Comment by pm47** [comment]
Actually went back to alpine for the final image. Best of both worlds or over the top?

**Comment by sethforprivacy** [comment]
Dockerfile looks GTG to me, fantasticly simplified approach compared to building from source.

I'll switch to using this as well in mine (will keep my repo just so there can be a published image in GHCR/Dockerhub) with proper credit, of course. Thanks for all of the work on this @pm47!

**Comment by sethforprivacy** [comment]
I have a few other minor improvements (none necessary, just tweaks and more explicit declaration of things like user:group pairings) in my PR on my own repo here, feel free to take anything that strikes your fancy, no credit etc. necessary.

https://github.com/sethforprivacy/phoenixd-docker/pull/3

**Comment by sethforprivacy** [comment]
Sorry for the message spam, but I actually just caught a breaking issue with the proposed Dockerfile. This Dockerfile doesn't properly set the file permissions for `/phoenix` which can sometimes cause the image to error out with a permissions issue when run.

See the fix here: https://github.com/sethforprivacy/phoenixd-docker/pull/3/commits/f3a63e0e2ca0fabda05bfa0573e3c9de730bfb0b

**Comment by sethforprivacy** [comment]
@pm47 seems there is a build issue for ARMv8, can't get it to find a workable Kotlin version. Logs:

https://paste.sethforprivacy.com/?ddbbbdc28345c06d#4DPdDZvXT4CGRBJ7jo1CDnNTUB3Ynkvp5yuMRcMJ3iqX

Not sure if you can access it, but the full Github Action run is here: https://github.com/sethforprivacy/phoenixd-docker/actions/runs/8457575444/job/23169856878?pr=3

**Comment by pm47** [comment]
> Edit: Nevermind, seems you can't build ARMv8 using ARMv8 due to Kotlin having no native aarch64 Linux support. Will find another way to distribute ARMv8 images then.

On the contrary, it should work, that is the point of using a jvm build for docker instead of a native one. Seems like gradle tries to build the native version, which we are not going to use anyway.

Will look at your other changes, feel free to suggest changes directly on this PR too!

File: .docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    && test `git rev-parse HEAD` = ${PHOENIXD_COMMIT_HASH} || exit 1 \
+    && ./gradlew distTar
+
+# Alpine image to minimize final image size
+FROM eclipse-temurin:21-jre-alpine
+
+# Upgrade all packages and install dependencies
+RUN apk update \
+    && apk upgrade --no-interactive
+RUN apk add --no-cache bash
```
Comment (sethforprivacy) [inline .docker/Dockerfile:27]: Prevent potential issues installing without also updating apk cache:

```suggestion
RUN apk add --update --no-cache bash
```

File: .docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+# Git pull phoenixd source at specified tag/branch and compile phoenixd
+WORKDIR /phoenixd
+RUN git clone --recursive --single-branch --branch ${PHOENIXD_BRANCH} -c advice.detachedHead=false \
+    https://github.com/ACINQ/phoenixd . \
+    && test `git rev-parse HEAD` = ${PHOENIXD_COMMIT_HASH} || exit 1 \
+    && ./gradlew distTar
+
+# Alpine image to minimize final image size
+FROM eclipse-temurin:21-jre-alpine
```
Comment (sethforprivacy) [inline .docker/Dockerfile:22]: Explicitly set label for final stage:

```suggestion
FROM eclipse-temurin:21-jre-alpine as FINAL
```

File: .docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+# Set necessary args and environment variables for building phoenixd
+ARG PHOENIXD_BRANCH=v0.1.3
+ARG PHOENIXD_COMMIT_HASH=d805f81c2bfb8a09a726bb36278216e607100a16
+
+# Upgrade all packages and install dependencies
+RUN apt-get update \
+    && apt-get upgrade -y
+RUN apt-get install -y --no-install-recommends bash git \
+    && apt clean
```
Comment (sethforprivacy) [inline .docker/Dockerfile:12]: Better apt cleanup:

```suggestion
    && apt clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```

File: .docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Ubuntu image for building, for compatibility with macOs arm64
```
Comment (sethforprivacy) [inline .docker/Dockerfile:1]:
```suggestion
# Use Ubuntu image for building for compatibility with macOS arm64 builds
```

File: .docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+RUN apk add --no-cache bash
+
+# Create a phoenix group and user
+RUN addgroup -S phoenix -g 1000 \
+    && adduser -S phoenix -G phoenix -u 1000 -h /phoenix
+USER phoenix
+
+# Unpack the release
+WORKDIR /phoenix
+COPY --chown=phoenix --from=BUILD /phoenixd/build/distributions/phoenix-*-jvm.tar .
```
Comment (sethforprivacy) [inline .docker/Dockerfile:36]: Explicitly set group in chown flag:

```suggestion
COPY --chown=phoenix:phoenix --from=BUILD /phoenixd/build/distributions/phoenix-*-jvm.tar .
```

**Review by sethforprivacy — CHANGES_REQUESTED** [review]
I think this encompasses all of my minor changes/improvements, feel free to take or leave any and all of them as you see fit :)

File: .docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+# Upgrade all packages and install dependencies
+RUN apk update \
+    && apk upgrade --no-interactive
+RUN apk add --no-cache bash
+
+# Create a phoenix group and user
+RUN addgroup -S phoenix -g 1000 \
+    && adduser -S phoenix -G phoenix -u 1000 -h /phoenix
+USER phoenix
```
[the remark this answers is not in the document]
Comment (pm47) [reply .docker/Dockerfile:32]: We don't need to create `.phoenix`, this would have hidden the bug fixed in #15.

**Comment by sstone** [comment]
Building `phoenixd` in docker is hard and there does not seem to be a way to make it work for `linux/arm64` (which would be a real production target), even if issues with gradle on alpine are solved.
Why not remove the build step and have the docker file simply download the released JVM binaries and use them directly ? that would give us a simple docker file that would work on all targets the JVM version supports (including `macos arm64/x64` and `linux arm64/x64`) ?

File: .docker/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: Understood, removed that in my Dockerfile now as well.

**Comment by sstone** [comment]
> Building `phoenixd` in docker is hard and there does not seem to be a way to make it work for `linux/arm64` 

@pm47 found a workaround (see #25)

File: .docker/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: Updated to only chown and not create any directories.

**Comment by sethforprivacy** [comment]
> @sethforprivacy the build should pass on arm at [75247e4](https://github.com/ACINQ/phoenixd/commit/75247e4de22ca5213c5012687975f0ccb25f205e) (current master tip)

Testing now! https://github.com/sethforprivacy/phoenixd-docker/actions/runs/8468177452/job/23200542411?pr=3

File: .docker/Dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: @pm47 when I don't create the directory explicitly I always get a failure when trying to run the Docker image in Github Actions:

```
Exception in thread "main" java.io.FileNotFoundException: /phoenix/.phoenix/seed.dat (Permission denied)
	at java.base/java.io.FileOutputStream.open0(Native Method)
	at java.base/java.io.FileOutputStream.open(Unknown Source)
	at java.base/java.io.FileOutputStream.<init>(Unknown Source)
	at okio.Okio__JvmOkioKt.sink(JvmOkio.kt:173)
	at okio.Okio.sink(Unknown Source)
	at okio.Okio__JvmOkioKt.sink$default(JvmOkio.kt:173)
	at okio.Okio.sink$default(Unknown Source)
	at okio.JvmSystemFileSystem.sink(JvmSystemFileSystem.kt:101)
	at fr.acinq.lightning.bin.conf.SeedKt.getOrGenerateSeed(Seed.kt:52)
	at fr.acinq.lightning.bin.Phoenixd$special$$inlined$defaultLazy$default$1.invoke(TransformAll.kt:253)
	at fr.acinq.lightning.bin.Phoenixd$special$$inlined$defaultLazy$default$1.invoke(TransformAll.kt:97)
	at com.github.ajalt.clikt.parameters.options.OptionWithValuesImpl.finalize(OptionWithValues.kt:221)
	at com.github.ajalt.clikt.internal.FinalizationKt.finalizeParameters(Finalization.kt:61)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:24[8](https://github.com/sethforprivacy/phoenixd-docker/actions/runs/8468177452/job/23200542411?pr=3#step:6:9))
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:41)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:457)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:454)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:474)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:481)
	at fr.acinq.lightning.bin.MainKt.main(Main.kt:72)
	```

This is fixed 100% of the time when I explicitly create + chown the default data directory. Are you sure that I shouldn't be doing that?

[the remark this answers is not in the document]
Comment (pm47) [reply .docker/Dockerfile:32]: If `phoenixd` is run by user `phoenix`, and user `phoenix` has recursive ownership of `/phoenix`, then `phoenixd` should be able to create dir `/phoenix/.phoenix`. We can hide the issue by creating the directory manually, but this is suspicious

[the remark this answers is not in the document]
Comment (sstone) [reply .docker/Dockerfile:32]: I think that the pb here is the ` VOLUME [ "/phoenix/.phoenix" ]` instruction which imho should not be needed (though I'm no docker expert).

[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: Testing in https://github.com/sethforprivacy/phoenixd-docker/pull/3/commits/b39e88b73721687342fd823eee94163e9c6f025b

[the remark this answers is not in the document]
Comment (pm47) [reply .docker/Dockerfile:32]: From https://serverfault.com/a/984599:

>Named volumes are initialized when first created to the contents of the image at the mount location. That initialization includes the owner and permissions. If /backup does not exist in your image, then an empty directory will be created and owned by root.

So in our case `/phoenix` is owned by `phoenix`, but `/phoenix/.phoenix` was owned by `root`. Your fix should work.

[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: You figured it out, @sstone :) Working perfectly now after changing the VOLUME directive, somehow that must have been creating it with bad permissions!

[the remark this answers is not in the document]
Comment (sethforprivacy) [reply .docker/Dockerfile:32]: Please reject this suggestion @pm47, no longer needed.

File: .docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+# Unpack the release
+WORKDIR /phoenix
+COPY --chown=phoenix:phoenix --from=BUILD /phoenixd/build/distributions/phoenix-*-jvm.tar .
+RUN tar --strip-components=1 -xvf phoenix-*-jvm.tar
+
+# Indicate that the container listens on port 9740
+EXPOSE 9740
+
+# Expose default data directory as VOLUME
+VOLUME [ "/phoenix/.phoenix" ]
```
Comment (sethforprivacy) [inline .docker/Dockerfile:43]: Fix permissions by only mounting the home directory:
```suggestion
VOLUME [ "/phoenix" ]
```

**Comment by pm47** [comment]
Looking pretty good I think. Will update branch/version after next release

**Comment by tsmith123** [comment]
Not sure if this Dockerfile is in its final state yet but it doesn't work for me on an M2 Macbook.

image

**Comment by sethforprivacy** [comment]
> Not sure if this Dockerfile is in its final state yet but it doesn't work for me on an M2 Macbook.
> 
> image

This issue was fixed in master, so next release will include the fix (per https://github.com/ACINQ/phoenixd/pull/17#issuecomment-2025186287). It was discussed at length above if you want more details on it!

**Comment by pm47** [comment]
@tsmith123 Please confirm you are getting this error when using the Dockerfile at commit [f255a11](https://github.com/ACINQ/phoenixd/pull/17/commits/f255a11d798bb13ea5f79374c1e7354c501cdb24).

**Comment by dnjooiopa** [comment]
> @tsmith123 Please confirm you are getting this error when using the Dockerfile at commit [f255a11](https://github.com/ACINQ/phoenixd/pull/17/commits/f255a11d798bb13ea5f79374c1e7354c501cdb24).

I got the same error on an M1 MacBook.
Commit f255a11d798bb13ea5f79374c1e7354c501cdb24
image

**Comment by pm47** [comment]
@tsmith123 @dnjooiopa Please try again with at [1c3a7ab](https://github.com/ACINQ/phoenixd/pull/17/commits/1c3a7ab5fea071b8987a20134d6bb1dcf44121a4) thanks!

**Comment by dnjooiopa** [comment]
> @tsmith123 @dnjooiopa Please try again at [1c3a7ab](https://github.com/ACINQ/phoenixd/pull/17/commits/1c3a7ab5fea071b8987a20134d6bb1dcf44121a4) thanks!

Great! Docker is now successfully building on my M1 Mac.

**Comment by sethforprivacy** [comment]
Amazing work on this, thank you @pm47! I've merged a very similar Dockerfile with automatic builds via Github Actions, so anyone who wants to use a ready-built image for phoenixd can use it easily:

https://github.com/sethforprivacy/phoenixd-docker/pkgs/container/phoenixd

The latest v0.1.4 version is building now and will be available shortly.

**Pull request merged**